### PR TITLE
React to changes of the deployed resources

### PR DESCRIPTION
**Pull request by mfaizanse**
**Description**

Changes proposed in this pull request:

- React to changes of the deployed resources

**Related issue(s)**

- https://github.com/kyma-project/nats-manager/issues/20

### CODE REVIEW

File: internal/controller/nats/controller.go [hunk 1]
```diff
@@ -25,26 +25,36 @@ import (
...
-	recorder      record.EventRecorder
-	logger        *zap.SugaredLogger
-	NATSManager   manager.Manager
+	controller.Controller
+	kubeClient                  k8s.Client
+	chartRenderer               chart.Renderer
+	Scheme                      *runtime.Scheme
+	recorder                    record.EventRecorder
+	logger                      *zap.SugaredLogger
+	NATSManager                 manager.Manager
```
Comment (marcobebway) [inline internal/controller/nats/controller.go:56]: Does it have to be exported?

File: internal/controller/nats/controller.go [hunk 2]
```diff
@@ -25,26 +25,36 @@ import (
...
-	kubeClient    k8s.Client
-	chartRenderer chart.Renderer
-	Scheme        *runtime.Scheme
-	recorder      record.EventRecorder
-	logger        *zap.SugaredLogger
-	NATSManager   manager.Manager
+	controller.Controller
+	kubeClient                  k8s.Client
+	chartRenderer               chart.Renderer
+	Scheme                      *runtime.Scheme
```
Comment (marcobebway) [inline internal/controller/nats/controller.go:53]: Does it have to be exported?

File: internal/controller/nats/controller.go [hunk 3]
```diff
@@ -25,26 +25,36 @@ import (
...
+	recorder                    record.EventRecorder
+	logger                      *zap.SugaredLogger
+	NATSManager                 manager.Manager
+	destinationRuleWatchStarted bool
```
Comment (marcobebway) [inline internal/controller/nats/controller.go:57]: Why the following properties are exported?

client.Client
controller.Controller
Scheme
NATSManager

File: pkg/k8s/types.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package k8s
+
+const (
+	// DestinationRuleCrdName defines the CRD name for DestinationRule as in Istio.
+	DestinationRuleCrdName string = "destinationrules.networking.istio.io"
+	// DestinationRuleKind defines the Kind name for DestinationRule as in Istio.
+	DestinationRuleKind string = "DestinationRule"
+	// DestinationRuleAPIVersion defines the API version for DestinationRule as in Istio.
+	DestinationRuleAPIVersion string = "networking.istio.io/v1alpha3"
```
Comment (marcobebway) [inline pkg/k8s/types.go:9]: Does it make sense to read these values from ENV config?

File: internal/controller/nats/provisioner.go [hunk 5]
```diff
@@ -38,7 +38,15 @@ func (r *Reconciler) handleNATSReconcile(ctx context.Context,
 		return ctrl.Result{}, r.syncNATSStatusWithErr(ctx, nats, err, log)
 	}
 
-	// @TODO: add watchers for deployed resources so if user modifies them, we should reconcile.
+	// watchers for dynamic resources managed by controller.
+	if instance.IstioEnabled && !r.destinationRuleWatchStarted {
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply internal/controller/nats/provisioner.go:42]: Its evaluated on every reconciliation.

File: pkg/k8s/types.go (hunk 4 above)
Comment (mfaizanse) [reply pkg/k8s/types.go:9]: I think, its not gonna be updated too frequently. We can keep it hardcoded here for now.

File: pkg/k8s/chart/releaseinstance.go [hunk 6]
```diff
@@ -9,17 +9,20 @@ import (
...
-func NewReleaseInstance(name, namespace string, configuration map[string]interface{}) *ReleaseInstance {
+func NewReleaseInstance(name, namespace string, istioEnabled bool,
+	configuration map[string]interface{}) *ReleaseInstance {
 	return &ReleaseInstance{
 		Name:          name,
 		Namespace:     namespace,
+		IstioEnabled:  istioEnabled,
 		Configuration: configuration,
 	}
 }
```
[the remark this answers is not in the document]
Comment (mfaizanse) [reply pkg/k8s/chart/releaseinstance.go:20]: Good catch, I updated it, its used [here](https://github.com/kyma-project/nats-manager/blob/d92645ed16be6626de5c3302d48e811329bedd85/internal/controller/nats/controller.go#L158) now.

File: internal/controller/nats/controller.go (hunk 3 above)
Comment (mfaizanse) [reply internal/controller/nats/controller.go]: I have update the following fields:
```
controller.Controller
Scheme
NATSManager
```

The `client.Client` is still exported so can refer to its methods from reconciler e.g `r.Create()`.

File: config/manager/kustomization.yaml [hunk 7]
```diff
@@ -11,3 +11,9 @@
 
 resources:
 - manager.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: controller
+  newName: mfaizan21/nats-manager-dev
+  newTag: "22030404"
```
Comment (marcobebway) [inline config/manager/kustomization.yaml:19]: Was this intentional?

Comment (mfaizanse) [reply config/manager/kustomization.yaml]: removed

**Review by marcobebway — APPROVED** [review]

**Pull request merged**